### ISSUE TITLE
new ast-check compile error: control flow implicitly returns from non-void function

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#10420]
Functions end with an implicit `return;` which implicitly returns a void value. Therefore, functions which return `void` have no requirement to have an explicit `return` expression:

```zig
fn foo(c: bool) void {
    if (c) {
        bar();
    }
}
```

However functions which do not return `void` require a `return` statement or there will be an error such as this:

```zig
fn foo(c: bool) i32 {
    if (c) {
        return 1234;
    }
}
```
```
./test.zig:1:21: error: expected type 'i32', found 'void'
fn foo(c: bool) i32 {
                    ^
```

Currently this error occurs during type checking, and so it won't trigger if the function is not referenced. It also will allow something like this:

```zig
const MyVoid = void;
fn foo(c: bool) MyVoid {
    if (c) {
        bar();
    }
}
```

Again, no error on this one.

This proposal is to add an additional check into `zig ast-check` (also known as `AstGen`). It makes the rules a bit stricter; functions which do not have a trivial `void` return type (literally the identifier `void` as the return type; no aliasing) are required to terminate the function body block with an expression of type `noreturn`. This could be `return` or it could be, for example, `unreachable`, or it could be a `while(true) {}` loop.

The above example (with `MyVoid`) would become:

```
./test.zig:6:1: error: control flow implicitly returns from non-void function
}
^
```

This compile error is directly related to all the `unreachable code` compile errors done by ast-check.

**Comment by ifreund** [comment]
I think we should amend this proposal to allow implicit returns in functions with return type `noreturn`. For example:
```zig
fn fatal() noreturn {
    std.log.err("something bad happened, can't recover\n", .{});
    std.os.exit(1);
}
```

**Comment by andrewrk** [comment]
Ah, the existence of `noreturn` functions complicates this a bit, will remove the accepted label to think about it some more. Without type-checking, zig must assume that any function *might* be a `noreturn` function.

**Comment by rohlem** [comment]
> the existence of `noreturn` functions complicates this a bit

Here's ~~a proposal~~ an idea: Require the user to explicitly mark `noreturn` function calls by following them with `unreachable`:

```zig
fn foo(c: bool) i32 {
    if (c) {
        return 1234;
    }
    exit(1);
    unreachable; // this clearly communicates intent, to both early compiler stages, and readers!
}
```

I think adequately solves this problem? Though honestly, I would really like this from a stylistic viewpoint alone.

**Comment by ifreund** [comment]
@rohlem then you don't get a compile error at the call site if, for example, the return type of `exit()` is changed to `void`.

**Comment by rohlem** [comment]
@ifreund Maybe we could introduce such a compile error?
To me `unreachable` sounds like the natural tool for signaling unreachable code.
IIUC, you would suggest something to the effect of `@compileErrorIfReachable()` (obviously with a better name)?

**Comment by ifreund** [comment]
> IIUC, you would suggest something to the effect of `@compileErrorIfReachable()` (obviously with a better name)?

We have this, it's called `@compileError()` :P

**Comment by ifreund** [comment]
> I'm naively hoping that evaluating its reachability analysis could be sufficiently postponed, don't know about the hurdles of implementing that.

Yeah, that's not how things work in practice. Unreachable code error MUST happen in AstGen (i.e. before semantic analysis) as branch at comptime allows arbitrary code to be only reachable when targeting a certain OS for example.

**Comment by rohlem** [comment]
> Unreachable code errors MUST happen in AstGen
@ifreund Why can't we emit a simple "unreachable poison" ZIR? (Special-case it if normal `unreachable` is not fit for this use case.)
Later in AIR, we know types and can determine whether to discard it (because we're in a dead code path) or error at this later stage.

I realize that this means the problem isn't isolated to be solved fully-verified during AstGen.
However, it would still allow the proposed ast-check to be implemented, with later validation whether the user-provided `unreachable` turned out true.
If it didn't, our function's missing a return value, so that's another problem we already need to deal with.
(If the user _wanted_ no return value, that should be written as `return undefined;` anyway.)

**Comment by rohlem** [comment]
> Ah, the existence of `noreturn` functions complicates this a bit, will remove the accepted label to think about it some more. Without type-checking, zig must assume that any function might be a `noreturn` function.
I believe this would be unblocked by https://github.com/ziglang/zig/issues/17234 .

**Comment by andrewrk** [comment]
This cannot be solved due to https://github.com/ziglang/zig/issues/10420#issuecomment-1001346503
